File: Server/src/main/java/cn/arorms/hsi/server/entities/HyperspectralImage.java
package cn.arorms.hsi.server.entities;

import cn.arorms.hsi.server.enums.ProcessStatus;
import com.fasterxml.jackson.annotation.JsonIncludeProperties;
import jakarta.persistence.*;
import lombok.Data;

import java.time.LocalDateTime;

/**
 * Entity representing a hyperspectral image.
 * 
 * @author Cacciatore
 * @version 1.0 2026-03-01
 */
@Data
@Entity
@Table(name = "hyperspectral_images")
public class HyperspectralImage {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String filename;

    @Enumerated(EnumType.STRING)
    private ProcessStatus status = ProcessStatus.PENDING;

    @Column(nullable = false, comment = "MAT file")
    private String matPath;

    @Column(comment = "BIN file")
    private String binPath;

    @Column(name = "pca_path", comment = "BIN file of PCA result")
    private String pcaPath;

    @Column(comment = "hash of file header 1MB")
    private String headerHash;

    @Column(comment = "JPG file for fake coloured image")
    private String overviewPicturePath;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "dataset_id")
//    @JsonIncludeProperties("dataset_id")
    private Dataset dataset;

    // Define by user
    @Column(name = "spatial_resolution", comment = "Length(m) of every pixel")
    private double spatialResolution;

    private Long fileSize;

    @Column(comment = "Data type of binary file")
    private String dataType;

    private LocalDateTime createdAt = LocalDateTime.now();
}

File: Server/src/main/kotlin/cn/arorms/hsi/server/entities/HyperspectralImage.kt
package cn.arorms.hsi.server.entities

import cn.arorms.hsi.server.enums.ProcessStatus
import jakarta.persistence.*

@Entity
@Table(name = "hyperspectral_images")
data class HyperspectralImage(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null,

    @Column(nullable = false)
    var filename: String = "",

    @Enumerated(EnumType.STRING)
    var status: ProcessStatus? = ProcessStatus.PENDING,

    // File name included
    @Column(nullable = false, comment = "MAT file")
    var matPath: String = "",
    @Column(comment = "BIN file")
    var binPath: String? = null,

    // Ready for file duplicated check
    @Column(comment = "hash of file header 1MB")
    var headerHash: String? = null,

    // TODO: Add a overview picture for fake coloured hsi display
    @Column(comment = "JPG file for fake coloured image")
    var overviewPicturePath: String? = null,

    // Metadata line from matlab file header
    var metadata: String? = null,

    // Size and shape
    var fileSize: Long? = null,
    var height: Int? = null,
    var width: Int? = null,
    var bands: Int? = null,

    ) {
    constructor(filename: String, matPath: String) : this() {
        this.filename = filename
        this.matPath = matPath
    }
}
